File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int dfs(TreeNode * root, int& num ){
        if(root==nullptr){
            return 0;
        }
        int leftheight= dfs(root->left,num);
        int rightheight= dfs(root->right,num);
        num= max(leftheight+rightheight,num);
        return max(leftheight,rightheight)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
       int num = 0;
       dfs(root,num);
        return num;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, prev_max):
            res = 0
            next_max = prev_max
            if node.val >= prev_max:
                res += 1
                next_max = node.val
            if node.left:
                res += dfs(node.left, next_max)
            if node.right:
                res += dfs(node.right, next_max)
            return res
        
        return dfs(root, root.val)